src/cudnn/conv.jl

```diff
@@ -10,9 +10,40 @@ using CUDA.CUDNN: scalingParameter, CUDNN_CONVOLUTION, convdims,
 
 const CUDNNFloat = Union{Float16,Float32,Float64}
 
-function cudnnConvolutionDescriptor(cdims::DenseConvDims, x::DenseCuArray{T}) where T
+function cudnnConvolutionDescriptorAndPaddedInput(cdims::DenseConvDims, x::DenseCuArray{T}) where T
+    # The main purpose of this function is to catch asymmetric padding which cudnn does not support
+    # If we find asymmetric padding we'll make a copy of x which is manually padded so that we can
+    # call cudnn with symmetric padding.
+    pad = collect(NNlib.padding(cdims)) # work with an array to make things more type stable
+    all(pad[1:2:end] .== pad[2:2:end]) && return (cudnnConvolutionDescriptor(cdims, x), x, identity)
+
+    # Maybe we should warn the user that this copies data, but other ML libs generally don't warn
+    sdims = NNlib.spatial_dims(cdims)
+    
+    # Naive implementation, is there a faster way?
+    # How much we need to pad x manually: The absolute difference between pad_left and pad_right, pad_top
+    # and pad_bottom etc. respectively. We keep the sign here though because we use it below to figure out
+    # which side of x to pad.
+    pad_manual = pad[1:2:2sdims] .- pad[2:2:2sdims]
+    # How much we can let cudnn pad: The smallest padding amount between pad_left and pad_right, pad_top 
+    # and pad_bottom etc. respectively
+    pad_cudnn = min.(pad[1:2:2sdims], pad[2:2:2sdims])
+
+
+    x_padded = similar(x, (size(x)[1:sdims] .+ abs.(pad_manual))..., size(x)[end-1:end]...)
```

Comment:
My understanding from https://github.com/FluxML/Flux.jl/pull/1759 is that slicing `Dims` as returned by `size` can cause instabilities. Worth a look.

Reply:
Ah great! I could have sworn I tried using `ntuple` and not getting anything more type stable. I changed all slicing/broadcasting to use `ntuple` and now `@code_warntype` gives all blue output. Benchmarks are a bit too noisy to tell if it really gave an improvement though which I guess is not too surprising.